### PR TITLE
feat: Add JsonSerializable(createJsonMeta)

### PR DESCRIPTION
**Pull request by rrousselGit**
As discussed before, this adds an option for generating a constant Map of `property name: json name`

fixes https://github.com/google/json_serializable.dart/issues/972

### CODE REVIEW

File: json_annotation/lib/src/json_serializable.dart [hunk 1]
```diff
@@ -64,6 +65,13 @@ class JsonSerializable {
   /// This setting has no effect if [createFactory] is `false`.
   final String? constructor;
 
+  /// If `true` (defaults to false), a private, static `_$ExampleJsonMeta`
+  /// constant is created in the generated part file.
+  ///
+  /// This constant can be used by other code-generators to support features
+  /// such as [fieldRename].
+  final bool? createJsonMeta;
```
Comment (kevmoo) [inline json_annotation/lib/src/json_serializable.dart:73]: "meta" is pretty vague here. Maybe `createFieldMap`? Let's talk before you start rewriting everything...

File: json_serializable/lib/src/encoder_helper.dart [hunk 2]
```diff
@@ -16,6 +16,26 @@ import 'unsupported_type_error.dart';
 abstract class EncodeHelper implements HelperCore {
   String _fieldAccess(FieldElement field) => '$_toJsonParamName.${field.name}';
 
+  /// Generates an object containing metadatas related to the encoding,
+  /// destined to be used by other code-generators.
+  String createJsonMeta(Set<FieldElement> accessibleFieldSet) {
+    assert(config.createJsonMeta);
+
+    final buffer =
+        StringBuffer('const _\$${element.name}JsonMeta = <String, String> {');
```
Comment (kevmoo) [inline json_serializable/lib/src/encoder_helper.dart:25]: Careful using `element.name` – if it's already private you'll end up with double `_` which makes analyzer sad.

I have a helper around here somewhere for this"

File: json_serializable/test/integration/json_meta_example.g.dart [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// **************************************************************************
+// JsonSerializableGenerator
+// **************************************************************************
+
+Model _$ModelFromJson(Map<String, dynamic> json) => Model(
+      firstName: json['first-name'] as String,
+      lastName: json['LAST_NAME'] as String,
+    );
+
+const _$ModelJsonMeta = <String, String>{
```
Comment (kevmoo) [inline json_serializable/test/integration/json_meta_example.g.dart:16]: Yeah. `_$ModelFieldMap` would be better here, right?

File: json_annotation/lib/src/json_serializable.dart (hunk 1 above)
Comment (rrousselGit) [reply json_annotation/lib/src/json_serializable.dart:73]: xFieldMap sounds good to me

**Comment by kevmoo** [comment]
still hacking on the failures @rrousselGit ?

**Comment by rrousselGit** [comment]
Indeed.

I don't quite understand why the CI says generated code is outdated. It isn't for me locally

File: json_annotation/pubspec.yaml [hunk 4]
```diff
@@ -12,6 +12,7 @@ dependencies:
   meta: ^1.4.0
 
 # When changing JsonSerializable class.
-# dev_dependencies:
-#  build_runner: ^2.0.0
-#  json_serializable: any
+dev_dependencies:
```
Comment (kevmoo) [inline json_annotation/pubspec.yaml:15]: You just have to hack this in on our local machine. Don't commit this!

**Comment by rrousselGit** [comment]
Looks like it's good now 😄

**Pull request merged**